### ISSUE TITLE
Intercept Screen

### ISSUE THREAD

**Issue by grill2010** [komefai/PS4RemotePlayInterceptor#2]
Hi, thank you for your awesome project. I'm the developer of DroidJoy a remote gamepad App for PC where you can emulate DInput and XInput gamepads. I constantly get feature requests to support consoles like the PS4 to emulate gamepads. I spent now a couple of days try to reveres engineer the remote play apk and the remote play application for PC. Unfortunately, as you know it is highly obfuscated and the major parts are in a native library. On Android the library is even encoded, although I managed to decode the library, every release Sony is changing something and you can start from the beginning.

I came to the conclusion that it would be easier to directly manipulate and intercept events from the running official remote play software, because it would be a more permanent solution I guess. As there is no official version for Android (only for Sony devices) I will stick to the remote play application for PCs. A couple of days ago I found your project by chance. And surprise, you already managed to simulate and inject events to the remote play software. Awesome!

The overall idea is, to develop a server which your Android Smartphone can connect to, so that you can more or less remote control the official remote play software by only using your Smartphone. Because I already developed a server for my current App DroidJoy, I have some experience in keeping the network delays between Smartphone and PC as small as possible. But I don't know if this will be sufficient because it always depends on the users local network quality. 

I don't know now if I should build my own interceptor or if I should stick to your one because it seems to be perfect. Furthermore, I think it should be possible to capture the screen of the PS4 by intercepting some events, you only have to find the part where you have to intercept. Do you have already tried something in this direction? If not, no problem. I will try to find out how this is possible.

**Comment by komefai** [comment]
Hi, thank you for your interest. I'm glad we came to the same conclusion of using the official software to control a PS4. This library has survived the update from PS4 firmware v4 to v5, so that is even more proof! Your app seems really cool and useful, I'm happy that you are considering integrating the library into your project.

For capturing screens, I've been using a simple helper function on [PS4 Macro](https://github.com/komefai/PS4Macro) which just takes the handle of the RemotePlay window and create a bitmap from it. You can find the code [here](https://github.com/komefai/PS4Macro/blob/master/PS4MacroAPI/Internal/BackgroundCapture.cs). I've used this method to make multiple game bots and it can easily capture at 30fps (at 1024x768 resolution) without taking too much resources.

However, since you are also passing it over the network, a better (more efficient) way would be to intercept it directly from RemotePlay as you've suggested. I haven't tried this method before, but I have a basic understanding of how they used native code to render the graphics. 

As far as I know, the C# winform app pass the handle of the .NET panel to the native DLL. Within the native code, it reads the compressed buffer from the network and somehow "stitches" them into a frame to render to the panel. I assumed that it "stitches" because of the distortion seen in [this video](https://www.youtube.com/watch?v=22ejWHlYebk) at the 8:05 mark.

I don't have much knowledge with DirectX or low level graphics, but I love to have the feature in this library if possible. But for now, no progress has been made in that area. You can make your own interceptor if this one doesn't fit your needs, but feel free to contribute to this project if you found a way.

Hope this helps!

**Comment by grill2010** [comment]
Sorry for the late response. Yep, that it survived the update from v4 to v5 seems to be a good sign.
Your solution of capturing the screen is really nice :) However, as you have already suggested, intercepting it directly from the PS4 remote play would be more efficient.

Thank you for your hints, I think I've found the part where it passes the handle to the native DLL. I will try to solve this problem. I have a few more other projects but as soon as I will find something I will let you know.

Kind regards

Ps.: I had not the time to dig in the code yet, but what is the difference between "Compatibility mode" and "Auto mode"?

**Comment by komefai** [comment]
No problem! If you need any help, do not hesitate to contact me :D

`InjectionMode.Compatibility` should only be used when `InjectionMode.Auto` causes an error about `System.Runtime.Remoting.Channels.Ipc.IpcChannel`. 

"Auto" will be used by default because it allows RemotePlay to be injected any number of times without having to restart it, but it is known to cause the above error on some Windows settings (probably related to user permissions/admin rights I guess). 
While "Compatibility" are the recommended way from [EasyHook's tutorial](http://easyhook.github.io/tutorials/remotefilemonitor.html) and are guaranteed to work on all Windows settings (hence the name), but the disadvantage of this method is that you can only inject ONCE into RemotePlay unless you restart it (as another PID).

In short, `InjectionMode.Compatibility` should work in all machines but has restrictions to the API.

**Comment by grill2010** [comment]
Hi, finally after almost 3 weeks I did a small step forward. I can finally intercept the screen from the PS4 Remote Play app by hooking into the DirectX calls. It was really a pain, especially because I had no DirectX experience. Nevertheless, the basic functionality is now working and I will try to optimize it in regard to the performance.

One big advantage is that it is now even working when the PS4 Remote Play app is minimized or covered by another application. I'm currently really busy because of daily work life and Christmas time. Next steps which I have in mind is to integrate the screen intercept feature into your library and I would also like to document my experience with the DirectX intercepting because maybe someone can improve it or has any suggestions of how we could make it better.

There were two major things which helped me to achieve this goal. First of all the open source DirectX-Hook project from @spazzarama and the program **API Monitor v2** for figuring out what is happening under the hood of the Remote Play app.

**Direct3DHook:**
https://github.com/spazzarama/Direct3DHook 
**API Monitor v2**
http://www.rohitab.com/apimonitor

**Comment by komefai** [comment]
Great work! Since Direct3DHook uses EasyHook, it should integrate nicely with this project without making big changes to the hooking logic. Documenting along the way is a good idea.

Looking forward to your contribution :)

**Comment by grill2010** [comment]
Hi, I released the source code of the PS4 Remote Play screen capturing program at

https://github.com/grill2010/Direct3DHook

I will build my own interceptor so if you want to include the screen capture functionality into your project feel free to use my code. In case you find any bugs just let me know.

**Comment by komefai** [comment]
Thanks for sharing your work! I'm going to close this issue now.
